File: packages/restapi/src/lib/chat/chats.ts
import { convertToValidDID, getAPIBaseUrls, isValidPushCAIP } from '../helpers';
import Constants, { ENV } from '../constants';
import { IFeeds } from '../types';
import {
  getInboxLists,
  addDeprecatedInfo,
  IPGPHelper,
  PGPHelper,
} from './helpers';
import { axiosGet } from '../utils/axiosUtil';
import { handleError } from '../errors/validationError';

export type ChatsOptionsType = {
  account: string;
  pgpPrivateKey?: string;
  /**
   * If true, the method will return decrypted message content in response
   */
  toDecrypt?: boolean;
  /**
   * page index - default 1
   */
  page?: number;
  /**
   * no of items per page - default 10 - max 30
   */
  limit?: number;
  /**
   * Environment variable
   */
  env?: ENV;
};

/**
 * Return the latest message from all wallet addresses you have talked to. This can be used when building the inbox page.
 */

export const chats = async (options: ChatsOptionsType): Promise<IFeeds[]> => {
  return await chatsCore(options, PGPHelper);
};

export const chatsCore = async (
  options: ChatsOptionsType,
  pgpHelper: IPGPHelper
): Promise<IFeeds[]> => {
  const {
    account,
    pgpPrivateKey,
    env = Constants.ENV.PROD,
    toDecrypt = false,
    page = 1,
    limit = 10,
  } = options || {};
  if (!isValidPushCAIP(account)) {
    throw new Error(`Invalid address!`);
  }
  const user = await convertToValidDID(account, env);
  const API_BASE_URL = getAPIBaseUrls(env);
  const apiEndpoint = `${API_BASE_URL}/v1/chat/users/${user}/chats?page=${page}&limit=${limit}`;
  const requestUrl = `${apiEndpoint}`;
  try {
    const response = await axiosGet(requestUrl);
    const chats: IFeeds[] = response.data.chats;
    const updatedChats = addDeprecatedInfo(chats);
    const feeds: IFeeds[] = await getInboxLists(
      {
        lists: updatedChats,
        user: user,
        toDecrypt,
        pgpPrivateKey,
        env,
      },
      pgpHelper
    );
    return feeds;
  } catch (err) {
    throw handleError(err, chats.name);
  }
};

File: packages/push-api-dart/lib/src/chat/src/chats.dart
import 'package:push_api_dart/push_api_dart.dart';

/// Return the latest message from all wallet addresses you have talked to.
/// This can be used when building the inbox page.
///
///toDecrypt: If true, the method will return decrypted message content in response
///page index - default 1
///limit: no of items per page - default 10 - max 30
Future<List<Feeds>?> chats({
  String? accountAddress,
  String? pgpPrivateKey,
  bool toDecrypt = false,
  int page = 1,
  int limit = 10,
}) async {
  String? userDID;
  if (accountAddress == null) {
    //copy cached did
    userDID = getCachedUser()?.did;
  } else {
    userDID = await getUserDID(address: accountAddress);
  }

  if (userDID == null) {
    throw Exception('Account address is required.');
  }

  pgpPrivateKey ??= getCachedUser()?.encryptedPrivateKey;
  if (pgpPrivateKey == null) {
    throw Exception('Private Key is required.');
  }

  try {
    final result = await http.get(
      path: '/v1/chat/users/$userDID/chats?page=$page&limit=$limit',
    );

    if (result == null || result['chats'] == null) {
      return null;
    }


    final chatList =
        (result['chats'] as List).map((e) => Feeds.fromJson(e)).toList();
    final updatedChats = chatList;
    addDeprecatedInfo(chatList);
    final feedWithInbox = await getInboxList(
      feedsList: updatedChats,
      user: userDID,
      pgpPrivateKey: pgpPrivateKey,
      toDecrypt: toDecrypt,
    );

    return feedWithInbox;
  } catch (e) {
    log(e);
    throw Exception('[Push SDK] - API chats: $e');
  }
}
